File: openvt/core/deform.py
from __future__ import annotations
import math
import numpy as np


def trs(pivot, tx, ty, rot_deg, sx, sy) -> np.ndarray:
    """M = T(pivot + t) * R * S * T(-pivot), 3x3 in canvas pixel space (y down)."""
    r = math.radians(rot_deg)
    c, s = math.cos(r), math.sin(r)
    a, b = c * sx, -s * sy
    d, e = s * sx, c * sy
    px, py = pivot
    return np.array([[a, b, px + tx - (a * px + b * py)],
                     [d, e, py + ty - (d * px + e * py)],
                     [0.0, 0.0, 1.0]])


def lerp_keys(keys, v):
    if v <= keys[0][0]:
        return keys[0][1]
    if v >= keys[-1][0]:
        return keys[-1][1]
    for i in range(len(keys) - 1):
        v0, a = keys[i]
        v1, b = keys[i + 1]
        if v0 <= v <= v1:
            t = 0.0 if v1 == v0 else (v - v0) / (v1 - v0)
            return a + (b - a) * t
    return keys[-1][1]


class DeformEngine:
    def __init__(self, model, params):
        self.model = model
        self.params = params
        self.nodes = model.nodes
        self.order = self._topo()
        self.tbind: dict[str, list] = {}
        self.mbind: dict[str, list] = {}
        for b in model.bindings:
            (self.tbind if b.type == "transform" else self.mbind).setdefault(b.target, []).append(b)
        self.world = {nid: np.eye(3) for nid in self.nodes}

    def _topo(self):
        depth = {}

        def d(nid, seen=()):
            if nid in depth:
                return depth[nid]
            n = self.nodes[nid]
            if n.parent is None or n.parent not in self.nodes or nid in seen:
                depth[nid] = 0
            else:
                depth[nid] = d(n.parent, seen + (nid,)) + 1
            return depth[nid]

        for nid in self.nodes:
            d(nid)
        return sorted(self.nodes, key=lambda k: depth[k])

    def evaluate(self):
        p = self.params
        for nid in self.order:
            n = self.nodes[nid]
            acc = np.array([0.0, 0.0, 0.0, 1.0, 1.0])
            for b in self.tbind.get(nid, ()):
                k = lerp_keys(b.keys, p.get(b.param))
                acc[:3] += k[:3]
                acc[3:] *= k[3:]
            local = trs(n.pivot, *acc)
            parent = self.world.get(n.parent) if n.parent is not None else None
            self.world[nid] = local if parent is None else parent @ local

        for layer in self.model.layers:
            pos = layer.base_pos
            mb = self.mbind.get(layer.id)
            if mb:
                pos = pos.copy()
                for b in mb:
                    pos += lerp_keys(b.keys, p.get(b.param))
            M = self.world[layer.id]
            layer.cur_pos = np.ascontiguousarray((pos @ M[:2, :2].T + M[:2, 2]).astype(np.float32))

File: openvt/web/js/deform.js
// 2x3 affine, row-major: [a,b,tx, d,e,ty]
export function trs(pivot, tx, ty, rot, sx, sy){
  const r = rot * Math.PI / 180, c = Math.cos(r), s = Math.sin(r);
  const a = c*sx, b = -s*sy, d = s*sx, e = c*sy, px = pivot[0], py = pivot[1];
  return [a, b, px + tx - (a*px + b*py), d, e, py + ty - (d*px + e*py)];
}
export function mul(A, B){
  return [
    A[0]*B[0] + A[1]*B[3], A[0]*B[1] + A[1]*B[4], A[0]*B[2] + A[1]*B[5] + A[2],
    A[3]*B[0] + A[4]*B[3], A[3]*B[1] + A[4]*B[4], A[3]*B[2] + A[4]*B[5] + A[5]
  ];
}
const IDENT = [1,0,0, 0,1,0];

function seg(keys, v){
  const n = keys.length;
  if (v <= keys[0].v) return [0, 0];
  if (v >= keys[n-1].v) return [n-1, 0];
  for (let i = 0; i < n-1; i++){
    const a = keys[i].v, b = keys[i+1].v;
    if (v >= a && v <= b) return [i, b === a ? 0 : (v - a) / (b - a)];
  }
  return [n-1, 0];
}

export class DeformEngine {
  constructor(model, params){
    this.model = model; this.params = params; this.nodes = model.nodes;
    this.tbind = new Map(); this.mbind = new Map();
    for (const b of model.bindings){
      const m = b.type === 'transform' ? this.tbind : this.mbind;
      if (!m.has(b.target)) m.set(b.target, []);
      m.get(b.target).push(b);
    }
    const depth = new Map();
    const d = (id, seen) => {
      if (depth.has(id)) return depth.get(id);
      const n = this.nodes.get(id);
      const v = (n.parent === null || !this.nodes.has(n.parent) || seen.has(id))
        ? 0 : d(n.parent, new Set([...seen, id])) + 1;
      depth.set(id, v); return v;
    };
    for (const id of this.nodes.keys()) d(id, new Set());
    this.order = [...this.nodes.keys()].sort((a, b) => depth.get(a) - depth.get(b));
    this.world = new Map();
  }

  evaluate(){
    const p = this.params;
    for (const id of this.order){
      const n = this.nodes.get(id);
      let tx = 0, ty = 0, rot = 0, sx = 1, sy = 1;
      for (const b of (this.tbind.get(id) || [])){
        const [i, t] = seg(b.keys, p.get(b.param));
        const A = b.keys[i].a, B = b.keys[Math.min(i+1, b.keys.length-1)].a;
        tx  += A[0] + (B[0]-A[0])*t;
        ty  += A[1] + (B[1]-A[1])*t;
        rot += A[2] + (B[2]-A[2])*t;
        sx  *= A[3] + (B[3]-A[3])*t;
        sy  *= A[4] + (B[4]-A[4])*t;
      }
      const local = trs(n.pivot, tx, ty, rot, sx, sy);
      const par = n.parent !== null ? this.world.get(n.parent) : null;
      this.world.set(id, par ? mul(par, local) : local);
    }

    for (const L of this.model.layers){
      const mb = this.mbind.get(L.id);
      let src = L.basePos;
      if (mb){
        const s = L.scratch; s.set(L.basePos);
        for (const b of mb){
          const [i, t] = seg(b.keys, p.get(b.param));
          const A = b.keys[i].off, B = b.keys[Math.min(i+1, b.keys.length-1)].off;
          for (let k = 0; k < s.length; k++) s[k] += A[k] + (B[k]-A[k])*t;
        }
        src = s;
      }
      const M = this.world.get(L.id) || IDENT, out = L.curPos;
      for (let k = 0; k < src.length; k += 2){
        const x = src[k], y = src[k+1];
        out[k]   = M[0]*x + M[1]*y + M[2];
        out[k+1] = M[3]*x + M[4]*y + M[5];
      }
    }
  }
}